src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -8916,6 +8918,1143 @@ public void NotExhaustive_LongList()
             Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithArguments("{ Length: 1000 }").WithLocation(2, 7));
     }
 
+    [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/60091")]
+    [InlineData("var")]
+    [InlineData("string[]")]
+    public void SlicePattern_DeclarationPatternOfSameTypeInBetween_Array_CodeGen(string typeSpecifier)
+    {
+        var source = $$"""
+            class C
+            {
+                string[] M(string[] s)
+                {
+                    if (s is ["(", ..{{typeSpecifier}} between, ")"])
+                    {
+                        return between;
+                    }
+
+                    return null;
+                }
+            }
+            """;
+
+        var comp = CreateCompilationWithIndexAndRangeAndSpan([source, TestSources.GetSubArray]);
+        comp.VerifyEmitDiagnostics();
```

Comment:
For tests that don't have errors, it would be good to verify execution.
I'd even recommend switching some of those tests away from arrays or ReadOnlySpans to a custom countable/indexable/sliceable type so that the order of evaluation can be made apparent. Verifying that the output is something like `index[0] index[^1] slice` will more effectively capture the intent and avoid unintentional regressions than a block of IL.

Reply:
> For tests that don't have errors, it would be good to verify execution.

Good idea to verify that behavior stays the same for most cases, working on that

> I'd even recommend switching some of those tests away from arrays or ReadOnlySpans to a custom countable/indexable/sliceable type so that the order of evaluation can be made apparent

There are 3 groups of tests added in this PR: when slice is in between other patterns, when it is the first and when it is the last. In each group there are 4 tests: 2 tests arrays with the same type and a narrowed type, test of a string, test of ROS and a test with custom side-effecting type. Each test has its own purpose: arrays and strings have special slicing behavior, so it is good to verify them separately, ROS represents a case when slicing is completely pure and test with side effect well... tests behavior change when there is a side effect (remember that this should represent a minority of real-world cases, so we shouldn't focus on behavior changes _too_ much). Thus I wouldn't switch existing tests from one type to another, rather add additional tests on top of what we already have.

> Verifying that the output is something like `index[0] index[^1] slice`

I didn't understand this one. Can you elaborate please? We already cover cases with side effects, what other types of side effects can we check for? Maybe we can modify existing side effect tests to verify more things at once? Can you please provide a prototype of a test you want to see